Mathlib/Tactic/LinearCombination.lean

```diff
@@ -119,25 +141,46 @@ theorem eq_of_add_pow [Ring α] [NoZeroDivisors α] (n : ℕ) (p : (a:α) = b)
 def elabLinearCombination
     (norm? : Option Syntax.Tactic) (exp? : Option Syntax.NumLit) (input : Option Syntax.Term)
     (twoGoals := false) : Tactic.TacticM Unit := Tactic.withMainContext do
-  let p ← match input with
-  | none => `(Eq.refl 0)
-  | some e => withSynthesize do
-    match ← expandLinearCombo e with
-    | none => `(Eq.refl $e)
-    | some p => pure p
+  let eqData := (← Tactic.getMainTarget).consumeMData.eq?.get!
+  let .sort u ← whnf (← inferType eqData.1) | unreachable!
+  let some v := u.dec | throwError "not a type"
+  let ((α : Q(Type v)), (a' : Q($α)), (b':Q($α))) := eqData
+  let (⟨a, b, p⟩ : Σ a b : Q($α), Q($a = $b)) ← do
+    match input with
+    | none =>
+      let _i ← synthInstanceQ q(OfNat ($α : Type v) 0)
+      pure ⟨q(0), q(0), q(Eq.refl 0)⟩
+    | some e => do
+      match ← expandLinearCombo α e with
+      | .const e => pure ⟨e, e, q(Eq.refl $e)⟩
+      | .proof a b p => pure ⟨a, b, p⟩
   let norm := norm?.getD (Unhygienic.run `(tactic| ring1))
-  Tactic.evalTactic <| ← withFreshMacroScope <|
-  if twoGoals then
-    `(tactic| (
-      refine eq_trans₃ $p ?a ?b
-      case' a => $norm:tactic
-      case' b => $norm:tactic))
-  else
-    match exp? with
-    | some n =>
-      if n.getNat = 1 then `(tactic| (refine eq_of_add $p ?a; case' a => $norm:tactic))
-      else `(tactic| (refine eq_of_add_pow $n $p ?a; case' a => $norm:tactic))
-    | _ => `(tactic| (refine eq_of_add $p ?a; case' a => $norm:tactic))
+  let (e, mvars) ← do
+    if twoGoals then
+      let mvar₁ ← mkFreshExprMVar q($a = $a')
+      let mvar₂ ← mkFreshExprMVar q($b = $b')
+      pure (q(eq_trans₃ (a' := $a') (b' := $b') $p), #[mvar₁, mvar₂])
+    else
+      match exp? with
+      | some n =>
+        let n : ℕ := n.getNat
+        if n = 1 then
+          let _i ← synthInstanceQ q(AddGroup ($α : Type v))
+          let mvar ← mkFreshExprMVar q($a' - $b' - ($a - $b) = 0)
+          pure (q(eq_of_add (a' := $a') (b' := $b') $p), #[mvar])
+        else
+          let _i ← synthInstanceQ q(Ring ($α : Type v))
+          let _i ← synthInstanceQ q(NoZeroDivisors ($α : Type v))
+          let mvar ← mkFreshExprMVar q(($a' - $b') ^ $n - ($a - $b) = 0)
+          pure (q(eq_of_add_pow (a' := $a') (b' := $b') $n $p), #[mvar])
+      | none =>
+        let _i ← synthInstanceQ q(AddGroup ($α : Type v))
+        let mvar ← mkFreshExprMVar q($a' - $b' - ($a - $b) = 0)
+        pure (q(eq_of_add (a' := $a') (b' := $b') $p), #[mvar])
+  (← Tactic.getMainGoal).assign (mkAppN e mvars)
+  for mvar in mvars do
+    Tactic.setGoals [mvar.mvarId!]
+    Tactic.evalTactic norm
```

Comment:
Should this assert that no goals remain? I think we might have a combinator for "run `tac` on `mvar` and make it close the goal"

Reply:
No, for example quite often one wants to run `linear_combination (norm := skip)`, precisely to see the remaining goals.

Reply:
Perhaps I'm missing something; how do these unsolved goals end up back in the tactic list? Doesn't `setGoals` leave only the last one around, forcing the user to `recover`?

Reply:
Ah, you're right: the tests don't contain an example in which simultaneously
- `mvars` contains more than one goal;
- `norm` does not close the first goal.

Here's an example like this:
```lean
example {a b c : ℤ} (h1 : a = 2) (h2 : b = c) : a + b = 2 + c := by
  linear_combination2 (norm := skip) h1 + h2
```
This indeed now produces an error "(kernel) declaration has metavariables '_example'".

Maybe you can tell me how to pick up the lost goal and add it back to the list -- I don't know how.

(By the way, the `linear_combination2` variant of the tactic is the only setting in which more than one goal would be generated, and as it happens, I've been planning to make a case for the deletion of that variant ....)

Reply:
OK, 5483c674ecf8e95b622117ea1765df150217d0ff is to address this, not sure if it's the idiomatic way but it seems to work.